django/contrib/admin/views/autocomplete.py

```diff
@@ -0,0 +1,167 @@
+        that is a small trade off.
+        """
+        return sys.maxsize
+
+
+class AutocompleteJsonView(BaseListView):
+    """
+    View that handles requests from `.AutocompleteWidget`.
+
+    The view only supports HTTP's GET method.
+    """
+    paginator_class = AutocompletePaginator
+    paginate_by = 20
+    admin_site = None
+
+    @property
+    def target_model(self):
+        return self.source_field.remote_field.target_model
+
+    @property
+    def source_field(self):
+        return self.source_model._meta.get_field(self.field_name)
+
+    @property
+    def target_model_admin(self):
+        try:
+            return self.admin_site._registry[self.target_model]
+        except KeyError:
+            raise Http404('%s is not registered in the admin.' % self.target_model)
+
+    @property
+    def source_model_admin(self):
+        return self.admin_site._registry[self.source_model]
+
+    def has_perm(self, request):
+        """Check if user has permission to access the related model."""
+        return self.source_model_admin.has_change_permission(request)
+
+    def get(self, request, *args, **kwargs):
+        """
+        Return JSON search result based on a field identifier and a search term.
+
+        Returns:
+             (django.http.JsonResponse)::
+                {
+                    results: [
+                        {
+                            id: "123"
+                            text: "foo",
+                        }
+                    ],
+                    pagination: {
+                        more: true
+                    }
+                }
+        """
+        from django.apps import apps
+
+        self.app_label = kwargs.get('app_label')
+        self.model_name = kwargs.get('model_name')
+        self.field_name = kwargs.get('field_name')
+        self.term = request.GET.get('term', '')
+        self.source_model = apps.get_model(self.app_label, self.model_name)
+
+        # Since the list of registered autocomplete fields is generated during the request
+        # this check must be performed in the view itself, rather in a restricted url pattern.
+        autocomplete_fields = self.source_model_admin.get_autocomplete_fields(request)
+        if self.field_name not in autocomplete_fields:
+            raise Http404('%s is not an autocomplete field.' % self.field_name)
+
+        if not self.has_perm(request):
+            raise PermissionDenied
+
+        self.object_list = self.get_queryset()
+        context = self.get_context_data()
+        return JsonResponse({
+            'results': [
+                {
+                    'id': force_text(obj.pk),
+                    'text': force_text(obj),
+                }
+                for obj in context['object_list']
+            ],
+            'pagination': {
+                'more': context['page_obj'].has_next()
+            },
+        })
+
+    def get_queryset(self):
```

Comment:
That is true. Since I now want to use the paginator form the `ModelAdmin` as well, it seems a good idea to just use the `ModelAdmin` all together.